### ISSUE TITLE
Raylet core dumped after RAY_CHECK failed in object manager

### ISSUE THREAD

**Issue by xutianming** [ray-project/ray#3741]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: CentOs 7
- **Ray installed from (source or binary)**: source 
- **Ray version**:0.6.1
- **Python version**:Anaconda 2.7.15
- **Exact command to reproduce**:
```python
import os
import ray
from ray.tune import run_experiments, grid_search


if __name__ == "__main__":
    env_dist = os.environ

    redis_address = env_dist.get("REDIS_ADDRESS")
    ray.init(redis_address=redis_address)

    run_experiments({
        "atari-impala-test": {
            "run": "IMPALA",
            "env": grid_search(["BreakoutNoFrameskip-v4",
                              "BeamRiderNoFrameskip-v4",
                              "QbertNoFrameskip-v4",
                              "SpaceInvadersNoFrameskip-v4"]),
            "stop": {
                "time_total_s": 200,
            },
            #"local_dir": "/data/bucket/",
            "config": {
                "sample_batch_size": 50,
                "train_batch_size": 500,
                "num_workers": 9,
                "num_envs_per_worker": 1,
                "clip_rewards": True,
            },
        },
    })
```

### Describe the problem

I was running above code with 3 nodes and raylet core dumped after RAY_CHECK failed.
The problem could not be reproduced with 1 node.
![image](https://user-images.githubusercontent.com/4970790/50971332-bd430900-151e-11e9-8732-ee89e6998740.png)
Above logs could be seen in raylet.err
The exception was thrown from [object_manager.cc#L612](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/object_manager.cc#L612)

### Source code / logs

I debugged and add some logs around L612.
```C++
void ObjectManager::SubscribeRemainingWaitObjects(const UniqueID &wait_id) {
  // ...
for (const auto &object_id : wait_state.object_id_order) {
    if (wait_state.remaining.count(object_id) > 0) {
      wait_state.requested_objects.insert(object_id);
      // Subscribe to object notifications.
      RAY_CHECK_OK(object_directory_->SubscribeObjectLocations(
          wait_id, object_id,
          [this, wait_id](const ObjectID &subscribe_object_id,
                          const std::unordered_set<ClientID> &client_ids, bool created) {
            if (!client_ids.empty()) {
              auto object_id_wait_state = active_wait_requests_.find(wait_id);
              if (object_id_wait_state == active_wait_requests_.end()) {
                // Depending on the timing of calls to the object directory, we
                // may get a subscription notification after the wait call has
                // already completed. If so, then don't process the
                // notification.
                return;
              }
              auto &wait_state = object_id_wait_state->second;
              RAY_LOG(ERROR) << "[tmxu] check object id: " << subscribe_object_id;
              if ( wait_state.remaining.count(subscribe_object_id) == 0 ) {
                RAY_LOG(ERROR) << "[tmxu] missing object id: " << subscribe_object_id;
              }

              RAY_CHECK(wait_state.remaining.erase(subscribe_object_id));
              RAY_LOG(ERROR) << "[tmxu] erased: " << subscribe_object_id;
              wait_state.found.insert(subscribe_object_id);
              wait_state.requested_objects.erase(subscribe_object_id);
              RAY_CHECK_OK(object_directory_->UnsubscribeObjectLocations(
                  wait_id, subscribe_object_id));
              if (wait_state.found.size() >= wait_state.num_required_objects) {
                WaitComplete(wait_id);
              }
            }
          }));
    }
  // ...
}
```
![image](https://user-images.githubusercontent.com/4970790/50971818-f7f97100-151f-11e9-9bd0-c8cb2d73ad12.png)

It seemed that one object id could be erased more than one time.
If so, we should remove RAY_CHECK at L612

**Comment by robertnishihara** [comment]
Thanks @xutianming for reporting this!

**Comment by stephanie-wang** [comment]
Closed by #3745.
